### PR TITLE
Expose the compiled Vocab file regex

### PR DESCRIPTION
**Pull request by mrm007**
... and expand it to match more extensions (`.js`, `.cjs`, `.mjs`)

### Other changes

`@vocab/core` now writes `import type` syntax in the generated translation file.
This syntax was introduced in TypeScript 3.8, so it should not be a breaking change.
This change goes hand in hand with https://github.com/SEEK-Jobs/metropolis/pull/2220 (and the [reasoning is the same](https://typescript-eslint.io/blog/consistent-type-imports-and-exports-why-and-how/))

`@vocab/core/compile.ts` was refactored to use `JSON.stringify` in many places that had custom logic to escape strings.
This should make the code easier to read and maintain.

`@vocab/webpack/loader` now generates a virtual module with the same export name as the translation file.
Vocab CLI generates `export default translations`.
But when Crackle/Rollup compiles the translation file, it will rename the export in the output file.
Then the export in the virtual module doesn't match any more.

### CODE REVIEW

File: packages/webpack/src/loader.ts [hunk 1]
```diff
@@ -103,14 +104,16 @@ export default async function vocabLoader(this: LoaderContext) {
...
   const result = /* ts */ `
     import { createLanguage, createTranslationFile } from '@vocab/webpack/${target}';
 
-    export default createTranslationFile({
+    const translation = createTranslationFile({
       ${loadedLanguages
         .map((lang) => `${JSON.stringify(lang)}: ${renderLanguageLoader(lang)}`)
         .join(',\n')}
     });
+    export { translation as ${exportName} };
```
Comment (askoufis) [inline packages/webpack/src/loader.ts:116]: This only affects the code resolved by the loader, the actual generated file hasn't changed.

File: .changeset/hip-deers-leave.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+'@vocab/webpack': minor
+---
+
+Expose the compiled Vocab file regex
```
Comment (askoufis) [inline .changeset/hip-deers-leave.md:5]: Should mention the name of the API here, `compiledVocabFileFilter`.

File: packages/webpack/src/loader.ts (hunk 1 above)
Comment (mrm007) [reply packages/webpack/src/loader.ts:116]: Correct:

https://github.com/seek-oss/vocab/blob/ce755f6140945460299593e27ec7b591ea1f0c22/packages/core/src/compile.ts#L193

File: packages/core/src/compile.ts [hunk 3]
```diff
@@ -124,30 +120,24 @@ function serialiseObjectToType(v: any) {
...
 
   for (const [key, value] of Object.entries(v)) {
-    if (value && typeof value === 'object') {
-      result += `'${encodeWithinSingleQuotes(key)}': ${serialiseObjectToType(
-        value,
-      )},`;
-    } else {
-      result += `'${encodeWithinSingleQuotes(key)}': ${value},`;
-    }
+    result += `${JSON.stringify(key)}: ${
```
Comment (jahredhope) [inline packages/core/src/compile.ts:123]: I think I understand this refactor, but it really confused me as it wasn't mentioned in the description and I wasn't sure why it was necessary.

Is this refactor relevant to the change?

File: packages/webpack/src/loader.ts (hunk 1 above)
Comment (askoufis) [reply packages/webpack/src/loader.ts:116]: Wasn't the point to affect the compiled file though, for ✨**rollup reasons**✨?

File: packages/core/src/compile.ts [hunk 4]
```diff
@@ -124,30 +120,24 @@ function serialiseObjectToType(v: any) {
...
 
-const banner = `// This file is automatically generated by Vocab.\n// To make changes update translation.json files directly.`;
-
-const serializeModuleImports = (imports: Set<string>, moduleName: string) => {
-  if (imports.size === 0) {
+const serializeTypeImports = (
+  imports: Set<string> | Array<string>,
+  moduleName: string,
+) => {
+  const importNames = Array.from(imports);
```
Comment (askoufis) [inline packages/core/src/compile.ts:135]: I don't see a change where `serializeTypeImports` is being passed an array, so IMO the type should stay as-is. Alternatively you could do `const importNames = new Set(imports);` to handle the array.

File: packages/core/src/compile.ts [hunk 5]
```diff
@@ -124,30 +120,24 @@ function serialiseObjectToType(v: any) {
...
+  if (importNames.length === 0) {
     return '';
   }
 
-  const importNames = Array.from(imports);
-
-  return `import { ${Array.from(importNames).join(
-    ', ',
-  )} } from '${moduleName}'`;
+  return `import type { ${importNames.join(', ')} } from '${moduleName}';`;
```
Comment (jahredhope) [inline packages/core/src/compile.ts:137]: Same as above question, is this relevant to the change or a separate improvement?

File: packages/webpack/src/loader.ts (hunk 1 above)
Comment (mrm007) [reply packages/webpack/src/loader.ts:116]: This is looking at the export in the compiled file.

Vocab CLI generates `export default translations`. When Crackle/Rollup compiles the translation file, it will rename the export in the output file. Then the export in the virtual module doesn't match any more.

File: packages/core/src/compile.ts [hunk 6]
```diff
@@ -124,30 +120,24 @@ function serialiseObjectToType(v: any) {
...
 
   return `{ ${result} }`;
 }
 
-const banner = `// This file is automatically generated by Vocab.\n// To make changes update translation.json files directly.`;
-
-const serializeModuleImports = (imports: Set<string>, moduleName: string) => {
-  if (imports.size === 0) {
+const serializeTypeImports = (
+  imports: Set<string> | Array<string>,
```
Comment (jahredhope) [inline packages/core/src/compile.ts]: Why does imports accept a Set or Array?

I can only see the function is called once with a set.

File: packages/core/src/compile.ts [hunk 7]
```diff
@@ -171,34 +161,37 @@ function serialiseTranslationRuntime(
       )}) => ${formatReturn}`;
     }
 
-    translationsType[encodeBackslash(key)] = translationFunctionString;
+    translationsType[key] = translationFunctionString;
```
Comment (jahredhope) [inline packages/core/src/compile.ts:161]: Was this previously not needed or what has changed to not need it now?

File: packages/core/src/compile.ts (hunk 4 above)
Comment (mrm007) [reply packages/core/src/compile.ts:135]: I had a version where `serializeTypeImports` was used more than once, but not any more.

File: packages/webpack/src/loader.ts [hunk 8]
```diff
@@ -103,14 +104,16 @@ export default async function vocabLoader(this: LoaderContext) {
 
   const loadedLanguages = Object.keys(loadedTranslation.languages);
 
+  const exportName = findExportNames(source)[0];
   const result = /* ts */ `
     import { createLanguage, createTranslationFile } from '@vocab/webpack/${target}';
 
-    export default createTranslationFile({
+    const translation = createTranslationFile({
```
Comment (jahredhope) [inline packages/webpack/src/loader.ts]: This seems to be a change to the generated code, but doesn't seem related to the PR description. Why this change?

**Review by jahredhope — COMMENTED** [review]
I think I have some questions that could be answered by more of a PR description. It seems like there a few extra changes in here.

File: packages/core/src/compile.ts (hunk 3 above)
Comment (mrm007) [reply packages/core/src/compile.ts:123]: These changes were done along with the import/export of translation files we discussed. Since all the tests are passing I left the changes in.

Happy to move these changes to a separate PR.

File: packages/webpack/src/loader.ts (hunk 8 above)
Comment (mrm007) [reply packages/webpack/src/loader.ts]: Explained in this thread https://github.com/seek-oss/vocab/pull/135#discussion_r1166289149

**Review by askoufis — APPROVED** [review]

File: packages/core/src/compile.ts (hunk 7 above)
Comment (mrm007) [reply packages/core/src/compile.ts:161]: It's now handled here: https://github.com/seek-oss/vocab/blob/d4fa3b0a1829862168e6f3099ce749b9bc2388e0/packages/core/src/compile.ts#L122-L126
